### PR TITLE
Added test and fix for cookie decoding issue

### PR DESCRIPTION
**Pull request by richardprice**
A new test to check if the cookie value is correctly decoded on reading
from Request.Cookies and a fix for the existing issue the test
highlights.

### CODE REVIEW

**Comment by thecodejunkie** [comment]
Moving this to 1.4 milestone. @richardprice could you rebase so we have a running version to discuss on? Thanks :heart:

**Comment by jchannon** [comment]
ping @richardprice

**Comment by richardprice** [comment]
Sorry, missed this first time round, will sort soon.

On 24 September 2015 at 14:46, Jonathan Channon <notifications@github.com>
wrote:

> ping @richardprice <https://github.com/richardprice>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NancyFx/Nancy/pull/1390#issuecomment-142931938>.
>



-- 
The most exciting phrase to hear in science, the one that heralds new
discoveries, is not 'Eureka!' but 'That's funny...' - Isaac Asimov

**Comment by khellang** [comment]
This is good to go now?

**Comment by richardprice** [comment]
Yup.

On 28 October 2015 at 19:50, Kristian Hellang <notifications@github.com>
wrote:

> This is good to go now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NancyFx/Nancy/pull/1390#issuecomment-151970210>.
>



-- 
The most exciting phrase to hear in science, the one that heralds new
discoveries, is not 'Eureka!' but 'That's funny...' - Isaac Asimov

File: src/Nancy.Tests.Functional/Tests/CookieTests.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    using System.Linq;
+    using Bootstrapper;
+    using ModelBinding;
+
+    using Nancy.Cookies;
+
+    using Testing;
+    using Xunit;
+
+    public class CookieTestsFixture
```
Comment (khellang) [inline src/Nancy.Tests.Functional/Tests/CookieTests.cs]: @richardprice Can you either 1) rename the class, or 2) rename the file? :smile:

File: src/Nancy.Tests/Unit/Diagnostics/CustomInteractiveDiagnosticsFixture.cs [hunk 2]
```diff
@@ -127,7 +127,7 @@ public void Should_return_main_page_with_valid_auth_cookie()
             // When
             var result = browser.Get(diagsConfig.Path + "/interactive/providers/", with =>
                 {
-                    with.Cookie(DiagsCookieName, this.GetSessionCookieValue("password"));
+                    with.Cookie(DiagsCookieName, Nancy.Helpers.HttpUtility.UrlEncode(this.GetSessionCookieValue("password")));
```
Comment (khellang) [inline src/Nancy.Tests/Unit/Diagnostics/CustomInteractiveDiagnosticsFixture.cs]: I'd rather add a `using Nancy.Helpers` (or `using Helpers` might even do it) and get rid of the full name for these.

**Comment by khellang** [comment]
:heart: Sorry it took almost 2 years :stuck_out_tongue_closed_eyes:

**Comment by khellang** [comment]
This is a breaking change.

If you manually decode cookies from the `Request.Cookies` dictionary, you should remove the decoding:

```diff
- var cookie = HttpUtility.UrlDecode(request.Cookies[cookieName]);
+ var cookie = request.Cookies[cookieName];
````

**Pull request merged**